diff --git a/CMakeLists.txt b/CMakeLists.txt
index 1bb6f52..5130d0d 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -4,6 +4,12 @@ project(libnfs LANGUAGES C VERSION 1.11.0)
 
 if(MSVC)
   set(CMAKE_DEBUG_POSTFIX "d")
+  add_definitions(-D_CRT_SECURE_NO_WARNINGS -D_CRT_NONSTDC_NO_WARNINGS)
+  add_definitions(-D_WIN32 -DMHD_W32DLL -D_USRDLL -DBUILDING_MHD_LIB)
+
+  if(CMAKE_SYSTEM_NAME STREQUAL "WindowsStore")
+    add_definitions(-D_WIN32_WINNT=0x0A00)
+  endif()
 endif()
 
 
diff --git a/win32/win32_compat.c b/win32/win32_compat.c
index a6876f4..8233ee7 100644
--- a/win32/win32_compat.c
+++ b/win32/win32_compat.c
@@ -186,11 +186,13 @@ int win32_gettimeofday(struct timeval *tv, struct timezone *tz)
  
   if (NULL != tz)
   {
+#if !defined(WINAPI_FAMILY) || (WINAPI_FAMILY == WINAPI_FAMILY_DESKTOP)
     if (!tzflag)
     {
       _tzset();
       tzflag++;
     }
+#endif
     tz->tz_minuteswest = _timezone / 60;
     tz->tz_dsttime = _daylight;
   }
-- 
2.13.2.windows.1

