--- /dev/null
+++ b/CMakeLists.txt
@@ -0,0 +1,15 @@
+cmake_minimum_required(VERSION 3.0)
+
+project(easyhook)
+
+add_custom_target(easyhook ALL COMMAND msbuild /m EasyHook.sln /p:Configuration=netfx4-Release /p:Platform=$(Platform) /target:EasyHookDll WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR})
+
+if(CMAKE_SIZEOF_VOID_P EQUAL 4)
+	install(FILES ${CMAKE_CURRENT_SOURCE_DIR}/Build/netfx4-Release/x64/EasyHook32.dll DESTINATION bin)
+	install(FILES ${CMAKE_CURRENT_SOURCE_DIR}/Public/easyhook.h DESTINATION include/easyhook)
+	install(FILES ${CMAKE_CURRENT_SOURCE_DIR}/Build/netfx4-Release/x64/EasyHook32.lib DESTINATION lib)
+else()
+	install(FILES ${CMAKE_CURRENT_SOURCE_DIR}/Build/netfx4-Release/x64/EasyHook64.dll DESTINATION bin)
+	install(FILES ${CMAKE_CURRENT_SOURCE_DIR}/Public/easyhook.h DESTINATION include/easyhook)
+	install(FILES ${CMAKE_CURRENT_SOURCE_DIR}/Build/netfx4-Release/x64/EasyHook64.lib DESTINATION lib)
+endif()
--- a/EasyHookDll/EasyHookDll.vcxproj
+++ b/EasyHookDll/EasyHookDll.vcxproj
@@ -1,5 +1,5 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
-<Project DefaultTargets="Build" ToolsVersion="12.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+<Project DefaultTargets="Build" ToolsVersion="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <ItemGroup Label="ProjectConfigurations">
     <ProjectConfiguration Include="netfx3.5-Debug|Win32">
       <Configuration>netfx3.5-Debug</Configuration>
@@ -44,49 +44,49 @@
     <ConfigurationType>DynamicLibrary</ConfigurationType>
     <CharacterSet>Unicode</CharacterSet>
     <WholeProgramOptimization>true</WholeProgramOptimization>
-    <PlatformToolset>v120</PlatformToolset>
+    <PlatformToolset>v140</PlatformToolset>
   </PropertyGroup>
   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='netfx4-Release|Win32'" Label="Configuration">
     <ConfigurationType>DynamicLibrary</ConfigurationType>
     <CharacterSet>Unicode</CharacterSet>
     <WholeProgramOptimization>true</WholeProgramOptimization>
-    <PlatformToolset>v120</PlatformToolset>
+    <PlatformToolset>v140</PlatformToolset>
   </PropertyGroup>
   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='netfx3.5-Debug|Win32'" Label="Configuration">
     <ConfigurationType>DynamicLibrary</ConfigurationType>
     <UseOfAtl>Static</UseOfAtl>
     <CharacterSet>Unicode</CharacterSet>
-    <PlatformToolset>v120</PlatformToolset>
+    <PlatformToolset>v140</PlatformToolset>
   </PropertyGroup>
   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='netfx4-Debug|Win32'" Label="Configuration">
     <ConfigurationType>DynamicLibrary</ConfigurationType>
     <UseOfAtl>Static</UseOfAtl>
     <CharacterSet>Unicode</CharacterSet>
-    <PlatformToolset>v120</PlatformToolset>
+    <PlatformToolset>v140</PlatformToolset>
   </PropertyGroup>
   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='netfx3.5-Release|x64'" Label="Configuration">
     <ConfigurationType>DynamicLibrary</ConfigurationType>
     <CharacterSet>Unicode</CharacterSet>
     <WholeProgramOptimization>true</WholeProgramOptimization>
-    <PlatformToolset>v120</PlatformToolset>
+    <PlatformToolset>v140</PlatformToolset>
   </PropertyGroup>
   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='netfx4-Release|x64'" Label="Configuration">
     <ConfigurationType>DynamicLibrary</ConfigurationType>
     <CharacterSet>Unicode</CharacterSet>
     <WholeProgramOptimization>true</WholeProgramOptimization>
-    <PlatformToolset>v120</PlatformToolset>
+    <PlatformToolset>v140</PlatformToolset>
   </PropertyGroup>
   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='netfx3.5-Debug|x64'" Label="Configuration">
     <ConfigurationType>DynamicLibrary</ConfigurationType>
     <UseOfAtl>Static</UseOfAtl>
     <CharacterSet>Unicode</CharacterSet>
-    <PlatformToolset>v120</PlatformToolset>
+    <PlatformToolset>v140</PlatformToolset>
   </PropertyGroup>
   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='netfx4-Debug|x64'" Label="Configuration">
     <ConfigurationType>DynamicLibrary</ConfigurationType>
     <UseOfAtl>Static</UseOfAtl>
     <CharacterSet>Unicode</CharacterSet>
-    <PlatformToolset>v120</PlatformToolset>
+    <PlatformToolset>v140</PlatformToolset>
   </PropertyGroup>
   <Import Project="$(VCTargetsPath)\Microsoft.Cpp.props" />
   <ImportGroup Label="ExtensionSettings">
--- a/Public/easyhook.h
+++ b/Public/easyhook.h
@@ -36,9 +36,15 @@
 
 #else
 
+#ifndef NTDDI_VERSION
     #define NTDDI_VERSION           NTDDI_WIN2KSP4
+#endif
+#ifndef _WIN32_WINNT
     #define _WIN32_WINNT            0x500
-    #define _WIN32_IE_              _WIN32_IE_WIN2KSP4
+#endif
+#ifndef _WIN32_IE_
+    #define _WIN32_IE_              _WIN32_IE_IE70
+#endif
 
     #include <windows.h>
     #include <winnt.h>
