diff --git a/expat/CMakeLists.txt b/expat/CMakeLists.txt
index 3417b69..7925ba1 100644
--- a/expat/CMakeLists.txt
+++ b/expat/CMakeLists.txt
@@ -490,7 +490,8 @@ endif(NOT WIN32)
 expat_install(TARGETS expat EXPORT expat
                       RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
                       LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
-                      ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR})
+                      ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+                      INCLUDES DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
 
 expat_install(FILES lib/expat.h lib/expat_external.h DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
 
@@ -807,13 +807,13 @@ expat_install(
         ${CMAKE_CURRENT_BINARY_DIR}/cmake/expat-config.cmake
         ${CMAKE_CURRENT_BINARY_DIR}/cmake/expat-config-version.cmake
     DESTINATION
-        ${CMAKE_INSTALL_LIBDIR}/cmake/expat-${PROJECT_VERSION}/
+        ${CMAKE_INSTALL_LIBDIR}/cmake/expat/
 )
 expat_install(
     EXPORT
         expat
     DESTINATION
-        ${CMAKE_INSTALL_LIBDIR}/cmake/expat-${PROJECT_VERSION}/
+        ${CMAKE_INSTALL_LIBDIR}/cmake/expat/
     NAMESPACE
         expat::
 )
